linux/mcompile/keymap/keymap.cpp

```diff
@@ -109,355 +472,330 @@ void Split_US_To_3D_Vector(v_str_3D &all_US,v_str_1D completeList) {
-  printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");
-  v_str_1D in {"a", "b", "m", "w", "x", "y", "z"};
-  std::string  out;
-  for( int i=0; i< (int) in.size()-1; i++) {
-    out = get_Other_Char_FromUS(in[i], All_Vector);
+KMX_WCHAR KMX_get_KeyValUnderlying_From_KeyValUS(v_dw_3D & All_Vector, KMX_DWORD VK_US) {
+  KMX_DWORD VK_underlying;
+  for( int i=0; i< (int)All_Vector[0].size()-1 ;i++) {
+    for( int j=1; j< (int)All_Vector[0][0].size();j++) {
+      if ( ( All_Vector[0][i][j] == VK_US ) ) {
+        VK_underlying = All_Vector[1][i][j];
+        return VK_underlying;
+      }
+    }
   }
+  return VK_US;
 }
 
+KMX_DWORD KMX_get_KeyCodeUnderlying_From_KeyCodeUS(GdkKeymap *keymap, v_dw_3D &All_Vector, KMX_DWORD KC_US, ShiftState ss, int caps) {
+  KMX_DWORD KC_underlying;
+  std::u16string u16str = convert_DeadkeyValues_To_U16str(KMX_get_KeyVal_From_KeyCode(keymap, KC_US, ss, caps));
 
-//--------------------------------------
-int main(gint argc, gchar *argv[])
-{
-  gdk_init(&argc, &argv);
-  GdkDisplay *display = gdk_display_get_default();
-  if (!display) {
-    printf("ERROR: can't get display\n");
-    return 1;
-  }
-  GdkKeymap *keymap = gdk_keymap_get_for_display(display);
-  if (!keymap) {
-    printf("ERROR: Can't get keymap\n");
-    gdk_display_close(display);
-    return 2;
+  for( int i=0; i< (int)All_Vector[1].size()-1 ;i++) {
+    for( int j=1; j< (int)All_Vector[1][0].size();j++) {
+      if ( ( All_Vector[1][i][j] == (KMX_DWORD)  *u16str.c_str() ) ) {
+        KC_underlying = All_Vector[1][i][0];
+        return KC_underlying;
+      }
+    }
   }
+  return KC_US;
+}
 
-  // write content of xkb_symbols to 3D Vector
-  // I assume we use Keyboard US basic as base
-  std::string US_language    = "us";
-  const char* text_us        = "xkb_symbols \"basic\"";
+UINT  KMX_get_KeyCodeUnderlying_From_VKUS(KMX_DWORD VirtualKeyUS) {
+  return (8 + USVirtualKeyToScanCode[VirtualKeyUS]);
+}
 
-  v_str_3D All_Vector;
-  write_US_ToVector(All_Vector,US_language, text_us);
-  //test(All_Vector);
+KMX_DWORD KMX_get_VKUS_From_KeyCodeUnderlying(KMX_DWORD keycode) {
+  if ( keycode >7)
+    return  (KMX_DWORD) ScanCodeToUSVirtualKey[keycode-8];
 
-  // add contents of other keyboard to vector
-  append_other_ToVector(All_Vector,keymap);
-  //test(All_Vector);
+  return 0;
+}
 
-  extract_difference(All_Vector);
-  //test_in_out(All_Vector);
+std::u16string CodePointToU16String(unsigned int codepoint) {
+  std::u16string str;
 
-  //print_simple_map_US(All_Vector,1);    // 1 = non-shift
-  //print_simple_map_Other(All_Vector,1); // 1 = non-shift
-  test_specific_Characters(All_Vector);
-  gdk_display_close(display);
+  if constexpr (sizeof(wchar_t) > 2) {
+      str = static_cast<char16_t>(codepoint);
+  }
+  else if (codepoint <= 0xFFFF) {
+      str = static_cast<char16_t>(codepoint);
+  }
+  else {
+      codepoint -= 0x10000;
```

Comment:
We might want to assert that `codepoint < 0x10FFFF` so that it catches our attention if we get an invalid codepoint. We won't get an invalid surrogate pair since we drop the illegal bits below in line 796, but I think it would still be good to get the developer's attention if this should happen. Also we might want to assert that this is a little endian system.


```suggestion
...
inline bool isLittleEndianSystem() {
  char16_t test = 0x0102;
  return (reinterpret_cast<char *>(&test))[0] == 0x02;
}
...
      assert(codepoint < 0x10FFFF);
      assert(isLittleEndianSystem());
      codepoint -= 0x10000;
```